addons/l10n_hu_edi/models/account_move.py

```diff
@@ -0,0 +1,1050 @@
+        we get the rate from the currency rates.
+        """
+        if self.currency_id.name == 'HUF':
+            return 1
+        if self.company_id.currency_id.name == 'HUF':
+            squared_amount_currency = sum(line.amount_currency ** 2 for line in self.invoice_line_ids)
+            squared_balance = sum(line.balance ** 2 for line in self.invoice_line_ids)
+            return math.sqrt(squared_balance / squared_amount_currency)
+        return self.env['res.currency']._get_conversion_rate(
+            from_currency=self.currency_id,
+            to_currency=self.env.ref('base.HUF'),
+            company=self.company_id,
+            date=self.invoice_date,
+        )
+
+    def _l10n_hu_edi_set_chain_index_and_line_number(self):
+        """ Set the l10n_hu_invoice_chain_index and l10n_hu_line_number fields. """
+        self.ensure_one()
+        base_invoice = self._l10n_hu_get_chain_base()
+        if base_invoice == self:
+            self.l10n_hu_invoice_chain_index = -1  # -1 indicates a base invoice (0 indicates the chain index was not set).
+            next_line_number = 1
+        else:
+            # Lock base invoice to prevent concurrent updates, ensuring sequence integrity.
+            base_invoice._l10n_hu_edi_acquire_lock()
+
+            prev_chain_invoices = base_invoice._l10n_hu_get_chain_invoices()
+            if not self.l10n_hu_invoice_chain_index:
+                last_chain_invoice = prev_chain_invoices.sorted(lambda m: m.l10n_hu_invoice_chain_index)[-1]
+                self.l10n_hu_invoice_chain_index = last_chain_invoice.l10n_hu_invoice_chain_index + 1 or 1
+            else:
+                last_chain_invoice = prev_chain_invoices.filtered(lambda m: m.l10n_hu_invoice_chain_index == (self.l10n_hu_invoice_chain_index - 1 or -1))
+
+            next_line_number = (max(last_chain_invoice.line_ids.mapped('l10n_hu_line_number')) or 0) + 1
+
+        # Set l10n_hu_line_number consecutively, first on product lines, then on rounding line
+        for line_number, line in enumerate(
+            self.line_ids.filtered(lambda l: l.display_type in ['product', 'rounding']).sorted(lambda l: l.display_type),
+            start=next_line_number,
+        ):
+            line.l10n_hu_line_number = line_number
+
+    @contextlib.contextmanager
+    def _l10n_hu_edi_acquire_lock(self):
+        """ Acquire a write lock on the invoices in self. On exit, commit to DB. """
+        if not self:
+            yield
+            return
+        try:
+            with self.env.cr.savepoint(flush=False):
+                self.env.cr.execute('SELECT * FROM account_move WHERE id = ANY(%s) FOR UPDATE NOWAIT', [self.ids])
+        except LockNotAvailable:
+            raise UserError(_('Could not acquire lock on invoices - is another user performing operations on them?')) from None
+        yield
+        if self.env['account.move.send']._can_commit():
+            self.env.cr.commit()
```

Comment:
What is the idea of doing a commit here?  Normally you lock until the end of the transaction.  (also, I see this method used without the with.

Reply:
I agree, it's messy. I'll remove the commit here and do an explicit commit in places where it's actually necessary (i.e. when we need to write a transaction ID we received from NAV)